smart-contracts/contracts/UnlockTestV2.sol

```diff
@@ -0,0 +1,179 @@
+ * 50% of GNP growth rate)
+ * NOTE: This smart contract is partially implemented for now until enough Locks are deployed and
+ * in the wild.
+ * The partial implementation includes the following features:
+ *  a. Keeping track of deployed locks
+ *  b. Keeping track of GNP
+ */
+
+import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
+import "./Lock.sol";
+
+
+contract UnlockTestV2 is Ownable {
+
+  /**
+   * The struct for a lock
+   * We use deployed to keep track of deployments.
+   * This is required because both totalSales and yieldedDiscountTokens are 0 when initialized,
+   * which would be the same values when the lock is not set.
+   */
+  struct LockBalances {
+    bool deployed;
+    uint totalSales; // This is in wei
+    uint yieldedDiscountTokens;
+  }
+
+  modifier onlyFromDeployedLock() {
+    require(locks[msg.sender].deployed, "Only from previously deployed locks");
+    _;
+  }
+
+  uint public grossNetworkProduct;
+
+  uint public totalDiscountGranted;
+
+  // We keep track of deployed locks to ensure that callers are all deployed locks.
+  mapping (address => LockBalances) public locks;
+
+  // Events
+  event NewLock(
+    address indexed lockOwner,
+    address indexed newLockAddress
+  );
+
+  bool internal initialized;
+
+  function initialize(
+    address _owner
+  )
+    public {
+      require(!initialized);
+      owner = _owner;
+      grossNetworkProduct = 0;
+      totalDiscountGranted = 0;
+      initialized = true;
+    }
+
+  /**
+  * @dev Create lock
+  * This deploys a lock for a creator. It also keeps track of the deployed lock.
+  */
+  function createLock(
+    Lock.KeyReleaseMechanisms _keyReleaseMechanism,
+    uint _expirationDuration,
+    uint _keyPrice,
+    uint _maxNumberOfKeys
+  )
+    public
+    returns (Lock lock)
+  {
+
+    // create lock
+    Lock newLock = new Lock(
+      msg.sender,
+      _keyReleaseMechanism,
+      _expirationDuration,
+      _keyPrice,
+      _maxNumberOfKeys
+    );
+
+    // Assign the new Lock
+    locks[address(newLock)] = LockBalances({
+      deployed: true,
+      totalSales: 0,
+      yieldedDiscountTokens: 0
+    });
+
+    // trigger event
+    emit NewLock(msg.sender, address(newLock));
+
+    // return the created lock
+    return newLock;
+  }
+
+  /**
+   * This function returns the discount available for a user, when purchasing a
+   * a key from a lock.
+   * This does not modify the state. It returns both the discount and the number of tokens
+   * consumed to grant that discount.
+   * TODO: actually implement this.
+   */
+  function computeAvailableDiscountFor(
+    address _purchaser, // solhint-disable-line no-unused-vars
+    uint _keyPrice // solhint-disable-line no-unused-vars
+  )
+    public
+    pure
+    returns (uint discount, uint tokens)
+  {
+    // an example modification
+    return (42, 42);
+  }
+
+  function testNewMethod()
```

Comment:
Ha! and that too!

Reply:
Thanks!

Reply:
I think we'll make it explicit when doing upgrades later by first committing a full identical copy of the smart contract and then doing a 2nd commit with only the changes. This should make things much easier to track down.